### ISSUE TITLE
eth_estimateGas returns "The execution failed due to an exception"

### ISSUE THREAD

**Issue by ThomasProust** [paritytech/parity-ethereum#10147]
_Before filing a new issue, please **provide the following information**._

- **Parity Ethereum version**: 2.1.10
- **Operating system**: Linux
- **Installation**: one-line installer
- **Fully synchronized**: yes
- **Network**: ethereum Ropsten
- **Restarted**: yes

Calling `eth_estimateGas` returns an exception error.

#### Example:

Call:
```
[2019-01-08 02:29:47.971] [DEBUG] worker - { method: 'eth_estimateGas',
  params:
   [ { from: '0x2c9cb2f3759d188d94b1de9928714ea25892f07d',
       to: '0x76b495bc78d435ddbadac8bb5957f3f141abfd98',
       value: '0x0',
       data: '0xa9059cbb000000000000000000000000593545d60f7be74efb6236e5a5a86f504d287a710000000000000000000000000000000000000000000010f0cf064dd592000000' } ],
  id: 1,
  jsonrpc: '2.0' }
```

Response:
```
[2019-01-08 02:29:47.984] [WARN] worker - { jsonrpc: '2.0',
  error:
   { code: -32016,
     message: 'The execution failed due to an exception.' },
  id: 1 }
```

_Your issue description goes here below. Try to include **actual** vs. **expected behavior** and **steps to reproduce** the issue._

**Comment by joshua-mir** [comment]
@ThomasProust do you see any additional information in your node's logs?

**Comment by ThomasProust** [comment]
@joshua-mir Nope, there is no information at all in the logs regarding that error

**Comment by niklasad1** [comment]
@ThomasProust I could not reproduce the behavior but I think https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/client/client.rs#L1544-#L1545 is the cause of this exception 

//cc @tomusdrw can you confirm?

It is a bummer that we don't log the what type of exception occurred. Perhaps maybe we could encapsulate the exception cause in `CallError::Exceptional`. However, it is really hard to investigate it further without logs

**Comment by Levino** [comment]
So what is a quickfix / workaround? Use an earlier version of parity?

**Comment by Levino** [comment]
BTW I get this when trying to call a contracts method on mainnet.

**Comment by tomusdrw** [comment]
@niklasad1 yeah, that's possible it's coming from there.

@Levino do you get this every time with all possible calls? It possibly signifies that the execution couldn't be completed (even if we provide a lot of gas).

**Comment by Levino** [comment]
It is one specific method and also does not happen all the time. Very hard to reproduce. Will try to consistently reproduce in Monday

**Comment by Levino** [comment]
Here is my node configuration:

```yaml
version: "3.3"
services:
  parity:
    image: parity/parity:v2.2.7
    ports:
      - 30303:30303
    command: >
               --ws-interface all
               --jsonrpc-interface all
               --base-path /root/parity
               --no-ancient-blocks
               --no-serve-light
               --max-peers 250
               --snapshot-peers 50
               --min-peers 50
               --mode active
               --tracing off
               --pruning fast
               --allow-ips public
               --db-compaction ssd
               --cache-size 4096
               --ws-origins all
               --ws-hosts all
               --auto-update none
    volumes:
      - parity:/root/parity
    working_dir: /root/parity
    entrypoint: /home/parity/bin/parity
    user: root 
```

**Comment by Levino** [comment]
I still encounter the same issue.

**Comment by Levino** [comment]
What does `Fallback to `BlockId::Latest` on the server actually mean? Anyone able to help?

**Comment by joshua-mir** [comment]
@Levino https://github.com/paritytech/parity-ethereum/blob/7c335e87642b848f61ab3ee7bbff5964b53d6f11/rpc/src/v1/impls/eth.rs#L652-L656

What's going on is while we're trying to calculate the nonce for the transaction, we first attempt to count pending transactions and if there is no value there we fall back to just counting in the latest known block - this is fairly unrelated to the issue you are experiencing here.

**Comment by Levino** [comment]
Okay thank you. I just discovered something which I think is the (unrelated) cause of my issue. I will clarify once I am sure about this.

**Comment by Levino** [comment]
My bad. It was an exception in the contracts code because the state of the contract was not as expected. Would have been much better if one would know what exactly happend instead of the cryptic message "Computer says no". I will delete my above comments to unclutter this issue.

**Comment by jmooo** [comment]
I've been running into this on Parity v2.2.8, and Geth v1.8.22 and it seems to fail when `to` is set to certain contracts but not others? For example:

BeerCoin: https://etherscan.io/address/0x74c1e4b8cae59269ec1d85d3d4f324396048f4ac
parity:
```
curl -H "Content-Type: application/json" --request POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"to":"0x74c1e4b8cae59269ec1d85d3d4f324396048f4ac","data":"0xa9059cbb0000000000000000000000007cb57b5a97eabe94205c07890be4c1ad31e486a80000000000000000000000000000000000000000000000000000000000000001"}],"id":34}' http://localhost:8889
{"jsonrpc":"2.0","result":"0x1f159","id":34}
```
geth:
```
curl -H "Content-Type: application/json" --request POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"to":"0x74c1e4b8cae59269ec1d85d3d4f324396048f4ac","data":"0xa9059cbb0000000000000000000000007cb57b5a97eabe94205c07890be4c1ad31e486a80000000000000000000000000000000000000000000000000000000000000001"}],"id":34}' http://localhost:8889
{"jsonrpc":"2.0","id":34,"result":"0x1f159"}
```

USDCoin: https://etherscan.io/address/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48
parity:
```
curl -H "Content-Type: application/json" --request POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"to":"0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48","data":"0xa9059cbb0000000000000000000000007cb57b5a97eabe94205c07890be4c1ad31e486a80000000000000000000000000000000000000000000000000000000000000001"}],"id":34}' http://localhost:8889
{"jsonrpc":"2.0","error":{"code":-32016,"message":"The execution failed due to an exception."},"id":34}
```
geth:
```
curl -H "Content-Type: application/json" --request POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"to":"0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48","data":"0xa9059cbb0000000000000000000000007cb57b5a97eabe94205c07890be4c1ad31e486a80000000000000000000000000000000000000000000000000000000000000001"}],"id":34}' http://localhost:8889
{"jsonrpc":"2.0","id":34,"error":{"code":-32000,"message":"gas required exceeds allowance or always failing transaction"}}
```

**Comment by ZumZoom** [comment]
@jmooo tx to BeerCoin does not fail because it does not have any requires to fail. Tx to USDCoin fails because you did not specify correct `from` address which would have at least 1 token to transfer. For example this works:

```
curl -H "Content-Type: application/json" --request POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"from": "0xA910f92ACdAf488fa6eF02174fb86208Ad7722ba","to":"0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48","data":"0xa9059cbb0000000000000000000000007cb57b5a97eabe94205c07890be4c1ad31e486a80000000000000000000000000000000000000000000000000000000000000001"}],"id":34}' http://localhost:8889
{"jsonrpc":"2.0","id":34,"result":"0x9af7"}
```

**Comment by jmooo** [comment]
You shouldn't have to specify a `from` to get an estimate should you? This is new behavior as Parity previously didn't fail when you didn't specify a from for estimates.

**Comment by ZumZoom** [comment]
Sometimes you surely should specify `from` to get an estimate. Imagine smart-contract that wastes 100k gas if your address starts with `0x000` and does not waste if it doesn't :)

In case of ERC20 tokens you should provide correct `from` so that `estimateGas` would be able to pass all the requires. In your example `from` is set to default 0, and some requires expectedly fail. And `estimateGas` says that such txn would fail.

**Comment by joshua-mir** [comment]
I feel like this issue can be closed (in favor of #10360), because it actually looks like this is a matter of not being verbose enough and not some kind of failure in performing the estimation on our end. Let me know if you disagree with that assessment.
